### PR TITLE
Allow beam search to re-use computations from previous steps

### PR DESCRIPTION
**Pull request by gcooper-isi**
See #16 

To test, run decoding with the beam size parameter set

### CODE REVIEW

File: megatron/text_generation_utils.py [hunk 1]
```diff
@@ -404,6 +406,17 @@ def sample_sequence_batch(model, context_tokens, context_lengths,
...
+                    # instances being processed, so that is the dimension that needs to be repeated beam_size times
+                    if layer_past is not None:
+                        layer_past = list(layer_past)
+                        for i in range(len(layer_past)):
+                            layer_past[i] = list(layer_past[i])
+                            for j in range(len(layer_past[i])):
+                                t = layer_past[i][j]
+                                t_size = t.size()
+                                layer_past[i][j] = t.view(t_size[0], t_size[1], -1).repeat(1, 1, args.beam_size)\
+                                                    .view(t_size[0], t_size[1] * args.beam_size, t_size[2], t_size[3])
```
Comment (DarrellBest) [inline megatron/text_generation_utils.py:419]: I wouldn't recommend using a nested for loop here. Perhaps itertool.product() would speed things up? https://docs.python.org/3/library/itertools.html#itertools.product

Comment (gcooper-isi) [reply megatron/text_generation_utils.py:419]: I did a little speed test to see if itertools.product would help. Here's my code for timing the nested for loop version:

```
import time
  
start = time.time()
for i in range(10000):
    for j in range(10000):
        pass
print(time.time() - start)
```

And for the itertools.product version:

```
import time
from itertools import product

start = time.time()
for i,j in product(range(10000), range(10000)):
    pass
print(time.time() - start)
```

The first script printed 2.91090607643 and the second printed 4.9702129364, so it doesn't look like itertools.product() would be faster in this case.

Comment (DarrellBest) [reply megatron/text_generation_utils.py:419]: Well, according to Tim Peters "flat is better than nested" so I try to live by this and find alternative solutions. You could write a generator function to handle this or many other ways.

Comment (gcooper-isi) [reply megatron/text_generation_utils.py:419]: The reason I wrote this code as nested `for` loops was to save memory. The whole `layer_past` object can be pretty big for large models. The way this code is currently written, the tensors in `layer_past` get replaced in-place one at a time, so the garbage collector should be able to throw away old tensors as it goes if it needs to (now that I think about it, adding an explicit `del` statement for each pass through the loop might not be a bad idea). I'm not sure of an alternative way to write this code while still accomplishing this goal, but I'm certainly open to suggestions.

**Comment by gcooper-isi** [comment]
@DarrellBest I've merged the latest from main into this branch, so if you could run this code with the latest version to verify, that would be helpful.

**Pull request merged**